### PR TITLE
Ropes: Issues 2 - 4

### PR DESCRIPTION
**Pull request by modethirteen**
I intended to create one branch per issue, but due to the manner in which I kicked-off this project, I had to clean up my git history to group related changes by git commit. As a result, the initial PR is a collection of Issues 2-4 (each commit is a single issue solution), with Issue 1 being the initial development environment setup.

### CODE REVIEW

**Comment by modethirteen** [comment]
[Issue 2: Save Image Link](https://trello.com/c/VoQW8ofZ/394-issue-2-save-image-link)

**Comment by modethirteen** [comment]
[Issue 3: Image Index](https://trello.com/c/YfSSuPdt/390-issue-3-image-index)

**Comment by modethirteen** [comment]
[Issue 4: Image Seeds](https://trello.com/c/m1y9wAM3/395-issue-4-image-seeds)

**Comment by modethirteen** [comment]
Merging so that the PR for Issue 5 does not include earlier commits.

File: test/controllers/images_controller_test.rb [hunk 1]
```diff
@@ -0,0 +1,46 @@
+require 'test_helper'
+
+class ImagesControllerTest < ActionDispatch::IntegrationTest
+
+  test 'can get image container for image react component' do
+    image = Image.create(url: 'https://example.com/xyzzy.png')
```
Comment (darreneng) [inline test/controllers/images_controller_test.rb:10]: minor but usually in tests we want to use `Image.create!` instead of `Image.create`. The former will raise an exception if it fails

File: app/models/image.rb [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+  def hostname
+    @hostname = URI.parse(self[:url]).host
+  rescue URI::InvalidURIError
+    'the interwebs'
+  end
+
+  # NOTE (andy.vaughn@appfolio.com, 20210303): pagination is descending, from most recent id to earliest
+  def next
+    Image.where('id < ?', id).order('id DESC').first || Image.last
```
Comment (darreneng) [inline app/models/image.rb:14]: normally I would suggest not ordering based on ids since they should be treated as opaque data types but this is outside the scope of this project

File: app/controllers/images_controller.rb [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+                                         }).serializable_hash.to_json
+      end
+    end
+  end
+
+  def new
+    @image = Image.new
+  end
+
+  def create
```
Comment (darreneng) [inline app/controllers/images_controller.rb:33]: could we add tests for this controller action?  just having one for each if/else branch should be fine

File: app/controllers/images_controller.rb [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+                                             html: images_url,
+                                             json: images_url(format: :json)
+                                           }
+                                         }).serializable_hash.to_json
+      end
+    end
+  end
+
+  def new
+    @image = Image.new
```
Comment (darreneng) [inline app/controllers/images_controller.rb:30]: could we add a controller test for this action? Doesn't need to be crazy

File: app/javascript/images/src/components/image.js [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+      );
+    }
+    return (
+      <div className="row">
+        <div className="col-md-3">
+          <div className="card mb-3 box-shadow">
+            <img
+              className="card-img-top"
+              src={image.attributes.url}
+              alt={`Hosted by ${image.attributes.hostname}`}
```
Comment (darreneng) [inline app/javascript/images/src/components/image.js:72]: nice job adding an `alt` attribute 👍

File: app/controllers/images_controller.rb [hunk 6]
```diff
@@ -1,7 +1,31 @@
...
   protect_from_forgery with: :exception
 
-  def index; end
+  def index
+    respond_to do |format|
+      # TODO (andyv.vaughn@appfolio.con, 20210303): grab tags from request and include in redirect
+      format.html { redirect_to controller: 'application', action: 'home', status: :found }
+      format.json {
+        # TODO (andy.vaughn@appfolio.com, 20210301): filter API results by incoming tag(s)
+        @images = Image.all.order('id DESC')
```
Comment (darreneng) [inline app/controllers/images_controller.rb:10]: For the index I would suggest ordering by the `created_at` timestamp since ids should be treated as an opaque data type

File: app/javascript/test/images/components/image.spec.js [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+      },
+      links: {
+        previous: null,
+        next: null
+      }
+    });
+    const wrapper = mount(<Image id={2} />);
+    /* NOTE (andy.vaughn@appfolio.com, 20210305): hacky workaround for async component mounting, do not replicate */
+    await wrapper.instance().componentDidMount();
+    wrapper.setProps();
```
Comment (darreneng) [inline app/javascript/test/images/components/image.spec.js:34]: I think we can use `wrapper.update()` here to update the UI?

File: app/javascript/images/src/components/image.js [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+      error: null,
+      isLoaded: false,
+      image: null
+    };
+  }
+
+  async componentDidMount() {
+    const { id } = this.props;
+    try {
+      const data = await api.getImage(id);
```
[the remark this answers is not in the document]
Comment (modethirteen) [reply app/javascript/images/src/components/image.js:39]: agreed, a fully implemented component needs to _react_ 🙄 to state changes

File: app/controllers/images_controller.rb [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+  def new
+    @image = Image.new
+  end
+
+  def create
+    @image = Image.new(params[:image].permit(:url))
+    if @image.save
+      # TODO (andy.vaughn@appfolio.com, 20210302): localize flash message
+      flash[:success] = 'Image successfully submitted.'
+      redirect_to @image
```
[the remark this answers is not in the document]
Comment (modethirteen) [reply app/controllers/images_controller.rb:38]: yea, I prefer that syntax - this was the suggestion of one the devs, but verbosity is always preferred.

File: app/javascript/api/http.js [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+    })
+    .catch(e => Promise.reject(e))
+    .then(json => json.data);
+}
+
+/**
+ * @param {String} url
+ * @returns {Promise<Object>} - JSON-unserialized API data
+ */
+export function get(url) {
```
[the remark this answers is not in the document]
Comment (modethirteen) [reply app/javascript/api/http.js:38]: right, makes sense - i didn't mind writing the boilerplate stuff, obviously common http handling is modularized in AppFolio's react environment

File: test/controllers/application_controller_test.rb [hunk 11]
```diff
@@ -1,8 +1,16 @@
 require 'test_helper'
-class ApplicationControllerTest < ActionDispatch::IntegrationTest
-  test 'should get home' do
+
+class ImagesControllerTest < ActionDispatch::IntegrationTest
```
[the remark this answers is not in the document]
Comment (modethirteen) [reply test/controllers/application_controller_test.rb:3]: yea, this is definitely a botched copy/paste job - i wouldn't expect any other convention

File: test/models/image_test.rb [hunk 12]
```diff
@@ -0,0 +1,97 @@
+require 'test_helper'
+
+class ImageTest < ActiveSupport::TestCase
+  test 'can save image if url starts with https' do
+    url = 'https://example.com/foo.png'
+    image = Image.new(url: url)
+    assert(image.valid?)
```
[the remark this answers is not in the document]
Comment (modethirteen) [reply test/models/image_test.rb:7]: 👍

File: app/controllers/images_controller.rb (hunk 3 above)
Comment (modethirteen) [reply app/controllers/images_controller.rb:33]: 👍

File: app/controllers/images_controller.rb (hunk 4 above)
Comment (modethirteen) [reply app/controllers/images_controller.rb:30]: 👍

File: app/controllers/images_controller.rb (hunk 6 above)
Comment (modethirteen) [reply app/controllers/images_controller.rb:10]: yea, agreed - this is probably just a great example of extreme laziness, and relying on the implementation detail that id's are auto-incremented 😄

File: app/javascript/test/images/components/image.spec.js (hunk 7 above)
Comment (modethirteen) [reply app/javascript/test/images/components/image.spec.js:34]: right, that should work - though tbh I hate this hack altogether

File: app/views/layouts/_header.html.erb [hunk 13]
```diff
@@ -0,0 +1,15 @@
+<header>
+  <nav class="navbar navbar-dark bg-dark">
+    <a class="navbar-brand" href="/">
+      <img class="d-inline-block align-middle" alt="<%= t('site_name') %>" src="<%= image_url('logo.png') %>" height="50px" />
+      <%= t('site_name') %>
+    </a>
+    <% unless current_page?(new_image_path) %>
```
[the remark this answers is not in the document]
Comment (modethirteen) [reply app/views/layouts/_header.html.erb:7]: oh right, I forgot about `yield` as a DI for templating.

File: app/views/layouts/_header.html.erb [hunk 14]
```diff
@@ -0,0 +1,15 @@
...
+      <form class="form-inline">
+        <a class="btn btn-primary" href="/images/new" role="button">
+          Submit an Image
+        </a>
+      </form>
```
[the remark this answers is not in the document]
Comment (modethirteen) [reply app/views/layouts/_header.html.erb:12]: no, this is just an oversight

File: test/controllers/images_controller_test.rb (hunk 1 above)
Comment (modethirteen) [reply test/controllers/images_controller_test.rb:10]: 👍

File: app/models/image.rb (hunk 2 above)
Comment (modethirteen) [reply app/models/image.rb:14]: 👍
